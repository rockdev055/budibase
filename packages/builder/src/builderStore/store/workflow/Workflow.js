import mustache from "mustache"
// TODO: tidy up import
import blockDefinitions from "../../../pages/[application]/workflow/WorkflowPanel/blockDefinitions"
import { generate } from "shortid"

/**
 * Class responsible for the traversing of the workflow definition.
 * Workflow definitions are stored in linked lists.
 */
export default class Workflow {
  constructor(workflow) {
    this.workflow = workflow
  }

  addBlock(block) {
    let node = this.workflow.definition
    while (node.next) node = node.next
    node.next = {
      id: generate(),
      ...block
    } 
  }

  updateBlock(updatedBlock, id) {
    let block = this.workflow.definition

    while (block.id !== id) block = block.next
    if (!block) throw new Error("Block not found.")

    block = updatedBlock
  }

  deleteBlock(id) {
    let previous = null
    let block = this.workflow.definition

    // iterate through the blocks
    while (block.id !== id) {
      previous = block
      block = block.next
    }

    // delete the block found
    previous.next = block.next || {}
  }

  createUiTree() {
    if (!this.workflow.definition.next) return []
    return Workflow.buildUiTree(this.workflow.definition.next)
  }

  static buildUiTree(block, tree = []) {
    if (!block) return tree

    // The client side display definition for the block
    const definition = blockDefinitions[block.type][block.actionId]
    if (!definition) {
      throw new Error(
        `No block definition exists for the chosen block. Check there's an entry in the block definitions for ${block.actionId}`
      )
    }

    if (!definition.params) {
      throw new Error(
        `Blocks should always have parameters. Ensure that the block definition is correct for ${block.actionId}`
      )
    }

    const tagline = definition.tagline || ""
    const args = block.args || {}

    tree.push({
      id: block.id,
      type: block.type,
      params: block.params,
      args,
      heading: definition.actionId,
      body: mustache.render(tagline, args),
    })

    return this.buildUiTree(block.next, tree)
  }
}
