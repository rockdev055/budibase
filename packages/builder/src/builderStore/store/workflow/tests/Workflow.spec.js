import Workflow from "../Workflow";
import TEST_WORKFLOW from "./testWorkflow";

const TEST_BLOCK = {
  id: "VFWeZcIPx",
  name: "Update UI State",
  tagline: "Update <b>{{path}}</b> to <b>{{value}}</b>",
  icon: "ri-refresh-line",
  description: "Update your User Interface with some data.",
  environment: "CLIENT",
  params: {
    path: "string",
    value: "longText",
  },
  args: {
    path: "foo",
    value: "started...",
  },
  actionId: "SET_STATE",
  type: "ACTION",
}

describe("Workflow Data Object", () => {
  let workflow

  beforeEach(() => {
    workflow = new Workflow({ ...TEST_WORKFLOW });
  });

  it("adds a workflow block to the workflow", () => {
    workflow.addBlock(TEST_BLOCK);
    expect(workflow.workflow.definition)
  })

  it("updates a workflow block with new attributes", () => {
    const firstBlock = workflow.workflow.definition.steps[0];
    const updatedBlock = {
      ...firstBlock,
      name: "UPDATED"
    };
    workflow.updateBlock(updatedBlock, firstBlock.id);
    expect(workflow.workflow.definition.steps[0]).toEqual(updatedBlock)
  })

  it("deletes a workflow block successfully", () => {
    const { steps } = workflow.workflow.definition
    const originalLength = steps.length 

    const lastBlock = steps[steps.length - 1];
    workflow.deleteBlock(lastBlock.id);
    expect(workflow.workflow.definition.steps.length).toBeLessThan(originalLength);
  })

  it("builds a tree that gets rendered in the flowchart builder", () => {
    expect(Workflow.buildUiTree(TEST_WORKFLOW.definition)).toMatchSnapshot();
  })
})
