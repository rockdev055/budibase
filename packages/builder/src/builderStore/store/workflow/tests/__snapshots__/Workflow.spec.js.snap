// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow Data Object builds a tree that gets rendered in the flowchart builder 1`] = `
Array [
  Object {
    "args": Object {
      "time": 3000,
    },
    "body": "Delay for <b>3000</b> milliseconds",
    "heading": "DELAY",
    "id": "zJQcZUgDS",
    "name": "Delay",
    "params": Object {
      "time": "number",
    },
    "type": "LOGIC",
  },
  Object {
    "args": Object {
      "path": "foo",
      "value": "finished",
    },
    "body": "Update <b>foo</b> to <b>finished</b>",
    "heading": "SET_STATE",
    "id": "3RSTO7BMB",
    "name": "Update UI State",
    "params": Object {
      "path": "string",
      "value": "longText",
    },
    "type": "ACTION",
  },
  Object {
    "args": Object {
      "path": "foo",
      "value": "started...",
    },
    "body": "Update <b>foo</b> to <b>started...</b>",
    "heading": "SET_STATE",
    "id": "VFWeZcIPx",
    "name": "Update UI State",
    "params": Object {
      "path": "string",
      "value": "longText",
    },
    "type": "ACTION",
  },
]
`;
